python/context.cpp

```diff
@@ -16,67 +18,158 @@
-             "Arguments:\n"
-             "  threads: The number of threads available locally for execution.\n"
-             "  gpu:     The index of the GPU to use, defaults to -1 for no GPU.\n")
-        .def_readwrite("threads", &arb::proc_allocation::num_threads,
+        .def(pybind11::init<int, pybind11::object>(),
+            "threads"_a=1, "gpu_id"_a=pybind11::none(),
+            "Construct an allocation with arguments:\n"
+            "  threads: The number of threads available locally for execution (defaults to 1).\n"
+            "  gpu_id:  The index of the GPU to use (defaults to None for no GPU).\n")
+        .def_property("threads", &proc_allocation_shim::get_num_threads, &proc_allocation_shim::set_num_threads,
             "The number of threads available locally for execution.")
-        .def_readwrite("gpu_id", &arb::proc_allocation::gpu_id,
+        .def_property("gpu_id", &proc_allocation_shim::get_gpu_id, &proc_allocation_shim::set_gpu_id,
             "The identifier of the GPU to use.\n"
             "Corresponds to the integer index used to identify GPUs in CUDA API calls.")
-        .def_property_readonly("has_gpu", &arb::proc_allocation::has_gpu,
+	.def_property_readonly("has_gpu", &proc_allocation_shim::has_gpu,
             "Whether a GPU is being used (True/False).")
-        .def("__str__", &proc_allocation_string)
-        .def("__repr__", &proc_allocation_string);
+        .def("__str__", &proc_alloc_string)
+        .def("__repr__", &proc_alloc_string);
 
-    pybind11::class_<context_shim> context(m, "context");
+// context
+    pybind11::class_<context_shim> context(m, "context", "An opaque handle for the hardware resources used in a simulation.");
     context
         .def(pybind11::init<>(
-            [](){return context_shim(arb::make_context());}))
+            [](){return context_shim(arb::make_context());}),
+            "Construct a local context with one thread, no GPU, no MPI by default.\n"
+            )
         .def(pybind11::init(
-            [](const arb::proc_allocation& alloc){return context_shim(arb::make_context(alloc));}),
+            [](const proc_allocation_shim& alloc){
+                return context_shim(make_context_shim(alloc));
+            }),
              "alloc"_a,
-             "Argument:\n"
+             "Construct a local context with argument:\n"
              "  alloc:   The computational resources to be used for the simulation.\n")
 #ifdef ARB_MPI_ENABLED
         .def(pybind11::init(
-            [](const arb::proc_allocation& alloc, mpi_comm_shim c){return context_shim(arb::make_context(alloc, c.comm));}),
-             "alloc"_a, "c"_a,
-             "Arguments:\n"
+            [](const proc_allocation_shim& alloc, pybind11::object mpi){
+                if (mpi.is_none()) {
+                    return context_shim(make_context_shim(alloc));
+                }
+                opt_mpi_comm c = py2optional<mpi_comm_shim>(mpi,
```

Comment:
You don't need the `opt_mpi_comm` type if you use `auto` here, i.e.
```C++
auto c = py2optional<mpi_comm_shim>(...);
```

Reply:
Done.